### PR TITLE
[IMP] portal: allow portal change login field

### PR DESCRIPTION
**Pull request by rahg-odoo**
Specifications:
  - Portal users should be able update their login ID through the
  "Edit Information" page on portal view.
    
Before this PR:
  - Only users with administrator rights could update
  login IDs for all users.
  - Portal users were restricted to changing their passwords
  but could not modify their login ID.
    
After this PR:
  - Portal users can update login ID via the "Edit Information"
    page on portal view.
    
task - 4269220

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/portal/controllers/portal.py [hunk 1]
```diff
@@ -202,7 +202,12 @@ def account(self, redirect=None, **post):
                         values[field] = False
                 values.update({'zip': values.pop('zipcode', '')})
                 self.on_account_update(values, partner)
+                login = values["login"]
+                del values["login"]
                 partner.sudo().write(values)
+                if request.env.user.login != login:
+                    request.env.user.write({"login": login})
+                    request.session.logout()
```
Comment (divy-odoo) [inline addons/portal/controllers/portal.py:210]: Not sure, but here we can sync with the change_password method by reloading the client instead logout.
image

Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:210]: @divy-odoo Thanks! we definitely need something to avoid that error or session does not expire first of all(best case scenario).

File: addons/portal/controllers/portal.py [hunk 2]
```diff
@@ -186,10 +186,15 @@ def account(self, redirect=None, **post):
         })
 
         if post and request.httprequest.method == 'POST':
+            login = post.get("login")
+            del post["login"]
```
[the remark this answers is not in the document]
Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:190]: Hi @msh-odoo  Form with ```id="portal_my_details"``` is the only one  sending post request to this controller. there will be always key named "login" in dict  of ```post```. value might be empty but login key will be always there so no need to verify it before deletion.

File: addons/portal/views/portal_templates.xml [hunk 3]
```diff
@@ -453,6 +453,10 @@
                 <br/>Please contact us directly for that operation.
             </small>
         </div>
+        <div t-attf-class="mb-3 #{error.get('login') and 'o_has_error' or ''} #{not login and 'd-none' or ''} col-xl-12">
```
[the remark this answers is not in the document]
Comment (rahg-odoo) [reply addons/portal/views/portal_templates.xml:456]: Hi @msh-odoo 
Hi @msh-odoo,  

We have two templates: `portal_my_details` (t1) and `ticket_validation_screen` (t2), both of which include the `portal_my_details_fields` (t3) template using `t-call`. The `t3` template contains a login field.

Both `t1` and `t2` interact with the controller routes `my/account` (r1) and `pos/ticket/validate` (r2) using `GET` and `POST` methods. However, when `r2` serves `t2` (which includes `t3`), the login field in `t3` will not be displayed in the response to a `GET` request from `r2`. This happens because the condition `#{not login and 'd-none' or ''}` ensures the login field is hidden when `login` is not present in the context of `r2`.

In summary:  
- The login field resides in `t3`.  
- Both `t1` and `t2` utilize `t3`.  
- We want only `t1` to display the login field, not `t2`. This is why the condition exists.

plot twist: i am not able to verify all of this, because `pos/ticket/validate` gives errors in master too. we have to discuss it with POS team.

[the remark this answers is not in the document]
Comment (msh-odoo) [reply addons/portal/views/portal_templates.xml:456]: @rahg-odoo We can do two improvements here: 1) Do not add field at all instead of adding ```d-none``` class 2) Instead of adding login field from Controller qcontext we can do the same thing from template from where we are calling template ```portal_my_details_fields```, you can call template using ```t-call``` and pass flag variable login with ```t-set``` variable

**Review by reth-odoo — COMMENTED** [review]
Why not, seems a bit odd that we don't ask for password to allow modifying the login but I guess that's just how it is.

Just a suggestion

**Comment by rahg-odoo** [comment]
> Why not, seems a bit odd that we don't ask for password to allow modifying the login but I guess that's just how it is.
> 
> Just a suggestion

@reth-odoo Nice idea! But to make in simple and make it similar to other fields we are not asking for password verification, just like instagram does not require any password verification for username updation. Certainly we could do that what do you say?

File: addons/portal/controllers/portal.py [hunk 4]
```diff
@@ -203,6 +208,10 @@ def account(self, redirect=None, **post):
                 values.update({'zip': values.pop('zipcode', '')})
                 self.on_account_update(values, partner)
                 partner.sudo().write(values)
+                if request.env.user.login != login:
+                    request.env.user.write({"login": login})
+                    # update session token so the user does not get logged out
+                    request.session.session_token = self.env.user._compute_session_token(request.session.sid)
```
[the remark this answers is not in the document]
Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:214]: @reth-odoo The condition `request.env.user.login != login` is not triggering an error. When the form details are loaded for a GET request at `my/account`, the `Login ID` field is pre-filled with the existing `login ID`. Therefore, if the user does not modify this field before submitting the form, no update is necessary.

[the remark this answers is not in the document]
Comment (reth-odoo) [reply addons/portal/controllers/portal.py:213]: Actually there's already a constraint for that, I was saying you can just catch validation errors on the write and make them look nice.

For the password check, was just a general note not necessarily in scope. Anyway we don't ask for password or confirmation to update the email either which is the more problematic of the two..

File: odoo/addons/base/models/res_users.py [hunk 5]
```diff
@@ -654,6 +654,18 @@ def _check_one_user_type(self):
             if self._has_multiple_groups(user_types_groups.ids):
                 raise ValidationError(_('The user cannot have more than one user types.'))
 
+    @api.constrains("login")
+    def _check_login(self):
```
Comment (kmagusiak) [inline odoo/addons/base/models/res_users.py:658]: Why is this needed? We alrady have a unique index on `login`. (check attribute `_login_key`).
Also the implementation checks the rowcount for the last execute.

File: addons/website/models/res_users.py [hunk 6]
```diff
@@ -21,7 +21,6 @@ class ResUsers(models.Model):
     @api.constrains('login', 'website_id')
     def _check_login(self):
```
Comment (kmagusiak) [inline addons/website/models/res_users.py:22]: Do it work today?
`_login_key` does already the job. Maybe review its definition.

**Review by kmagusiak — CHANGES_REQUESTED** [review]

File: addons/portal/controllers/portal.py [hunk 7]
```diff
@@ -184,15 +208,31 @@ def account(self, redirect=None, **post):
...
+            if not login:
+                error["login"] = "missing"
+                error_message.append(_("Login ID can not be empty!"))
             values.update({'error': error, 'error_message': error_message})
             values.update(post)
             if not error:
+                if request.env.user.login != login:
+                    try:
+                        request.env.user.write({"login": login})
+                        self.env.cr.commit()
```
Comment (kmagusiak) [inline addons/portal/controllers/portal.py:227]: Why don't use use a savepoint instead of commit?

Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:227]: @kmagusiak The main reason I used commit() is to rely on the database's unique constraint for the login field. When the transaction is committed, the database checks the constraint and raises an error if the login is not unique. If this happens, rollback() will undo the changes.

I could have used a savepoint to achieve the same result, but I chose not to because this is a single operation, not a series of steps. In my view, there’s no real difference between committing and rolling back later if the transaction fails, versus using a savepoint and rolling back to it. Both methods work fine, but using commit keeps things simpler in this case.

I can change according to what we have used at other places in codebase if you want.

Comment (kmagusiak) [reply addons/portal/controllers/portal.py:227]: The constraints are also checked at flush time (the ORM flushes by default when starting/ending a savepoint). If you have another exception after your commit, the transaction will rollback only up to the change of the login and might retry to do the whole call where you have already partially committed something. In general it is harder to prove it always works and if you have someone overwriting your method (think of a parter or a community addons), they may get a lot of side-effects. In your case, it might work, but savepoints are more advisable to use than commit/rollback.

In general, you should avoid to try to manage transactions in addons (commits, rollbacks, new cursors) whenever possible: that is the job of the core/ORM. 1 controller call should be 1 transaction.

Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:227]: @kmagusiak I misunderstood in first message. you can have a look in recent changes. Thanks for the review and insights!

File: addons/portal/controllers/portal.py [hunk 8]
```diff
@@ -185,14 +209,30 @@ def account(self, redirect=None, **post):
...
             values.update({'error': error, 'error_message': error_message})
             values.update(post)
             if not error:
+                if request.env.user.login != login:
+                    try:
+                        with request.env.cr.savepoint():
+                            request.env.user.write({"login": login})
+                        # update session token so the user does not get logged out
+                        request.session.session_token = request.env.user._compute_session_token(request.session.sid)
+                    except (ValidationError, psycopg2.errors.UniqueViolation) as e:
```
Comment (reth-odoo) [inline addons/portal/controllers/portal.py:231]: So do we actually ever get ValidationError in this case? It seems maybe a bit broad when the error message is much more specific

File: addons/portal/controllers/portal.py [hunk 9]
```diff
@@ -185,14 +209,30 @@ def account(self, redirect=None, **post):
...
+                        error["login"] = "Duplicate"
+                        error_message.append(_("You can not have two users with the same login."))
+                        values.update({'error': error, 'error_message': error_message})
+                        return self._prepare_my_details_values_and_return(values, partner, redirect, login)
```
Comment (reth-odoo) [inline addons/portal/controllers/portal.py:235]: Not sure why login error deserved an early return but not the other types of errors. If login is popped from the values it should be fine to continue no?

Comment (reth-odoo) [inline addons/portal/controllers/portal.py:235]: can't you do the login check separately, add the error if any and continue the logic as before?

File: addons/portal/controllers/portal.py (hunk 8 above)
Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:231]: Yes, we do get validationError.
context: we have access to my/account URL even without website if we have installed portal module.

'website/models/res_users.py' has methods named '_check_login' which throws 'ValidationError', but when website is not installed 'res_users.py' file from 'base' does not have any '_check_login' method so it uses exception from postgres converted to python exception using 'psycopg2' that is 'psycopg2.errors.UniqueViolation'.

when website is installed.
val

when website is not installed.
uni

IMO, we don't need different message for both of them as both exception is cause of single event that is duplicate login ID.

File: addons/portal/controllers/portal.py (hunk 9 above)
Comment (rahg-odoo) [reply addons/portal/controllers/portal.py]: @reth-odoo pushed changes. Thank you for review and insights!

**Comment by SergeBayet** [comment]
@rahg-odoo Could you check this weird alignment of the alert box?
image

**Review by SergeBayet — CHANGES_REQUESTED** [review]
Hi @rahg-odoo. Thanks for this PR. Here's my review. Don't forget to rebase and resolve your conflicts. Have a nice day! :+1:

**Comment by rahg-odoo** [comment]
@SergeBayet Thanks for the review! Unfortunately merge conflict is too big that I need to handle this task again. but i'll keep your review points in mind.

**Review by SergeBayet — CHANGES_REQUESTED** [review]

**Comment by SergeBayet** [comment]
> @SergeBayet Thanks for the review! Unfortunately merge conflict is too big that I need to handle this task again. but i'll keep your review points in mind.

Sorry to read that. Good luck with this one :+1:

**Review by kmagusiak — APPROVED** [review]

**Comment by rahg-odoo** [comment]
@msh-odoo made few changes after conflict resolution. Can please review once again? Thanks!

File: addons/portal/controllers/portal.py [hunk 10]
```diff
@@ -154,6 +155,20 @@ def _prepare_portal_layout_values(self):
...
         }
 
+    def _validate_login_field(self, login, invalid_fields, missing_fields, error_messages):
+        if not login:
+            missing_fields.add("login")
+            error_messages.append(_("Some required fields are empty."))
+        elif request.env.user.login != login:
+            try:
+                with request.env.cr.savepoint():
+                    request.env.user.write({"login": login})
```
Comment (msh-odoo) [inline addons/portal/controllers/portal.py:165]: Method name is related to validate login field but we are doing write operation, either method name should ```_validate_login_field_and_update``` or something like that, what do you say?

Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:165]: Yes, It is good idea. just changed that.

File: addons/portal/static/src/js/address.js [hunk 11]
```diff
@@ -198,7 +199,8 @@ publicWidget.registry.customerAddress = publicWidget.Widget.extend({
                 result.invalid_fields.forEach(
                     fieldName => this.addressForm[fieldName].classList.add('is-invalid')
                 );
-
```
Comment (SergeBayet) [inline addons/portal/static/src/js/address.js:201]: Unnecessary diff.

**Review by SergeBayet — CHANGES_REQUESTED** [review]
A few more remarks. See below.

File: addons/portal/controllers/portal.py [hunk 12]
```diff
@@ -496,6 +514,9 @@ def _create_or_update_address(
         """
         use_delivery_as_billing = str2bool(use_delivery_as_billing or 'false')
 
+        login = form_data.pop("login", False)
+        login = login.strip()
```
Comment (SergeBayet) [inline addons/portal/controllers/portal.py:518]: Did you test that? If login is `False`, I doubt that `False.strip()` is valid :)
Maybe use `form_data.pop("login", "").strip()` and later on, `if login` instead of `if login != False`

**Review by SergeBayet — CHANGES_REQUESTED** [review]
We're almost there.

File: addons/portal/controllers/portal.py (hunk 12 above)
Comment (SergeBayet) [reply addons/portal/controllers/portal.py:518]: Here it is:
image

Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:518]: @SergeBayet oh thanks! that was obvious.

**Comment by SergeBayet** [comment]
Be aware of that: https://github.com/odoo/odoo/pull/192098. I don't know when will it be merged though.

File: addons/portal/views/address_templates.xml [hunk 13]
```diff
@@ -134,6 +134,18 @@
     </template>
 
     <template id="portal.address_form_fields" name="Address Details">
+        <t t-if="login_field">
```
Comment (SergeBayet) [inline addons/portal/views/address_templates.xml:137]: `login_field` is always True or am I missing something?

File: addons/portal/views/address_templates.xml [hunk 14]
```diff
@@ -356,16 +368,16 @@
...
             </a>
-            <div class="position-relative w-100 d-flex d-md-none justify-content-center align-items-center order-2 my-2 opacity-75">
+            <div class="position-relative w-100 d-flex d-md-none justify-content-center align-items-center order-3 order-md-2 my-2 opacity-75">
                 <hr class="w-100"/>
                 <span class="px-3">or</span>
                 <hr class="w-100"/>
             </div>
-            <button id="save_address" class="btn btn-primary w-100 w-md-auto order-1 order-md-3">
+            <button id="save_address" class="btn btn-primary w-100 w-md-auto order-1 order-md-3 ms-auto gap-2">
                 <t name="save_address_label">Save Address</t>
```
Comment (SergeBayet) [inline addons/portal/views/address_templates.xml:381]:
```suggestion
                <t name="save_address_label">Save</t>
```

**Review by SergeBayet — CHANGES_REQUESTED** [review]
Hi @rahg-odoo! Here's some remarks. Cheers!

File: addons/portal/views/address_templates.xml (hunk 14 above)
Comment (rahg-odoo) [reply addons/portal/views/address_templates.xml:381]: @SergeBayet that has been adopted in another task. #185085

Comment (SergeBayet) [reply addons/portal/views/address_templates.xml:381]: > @SergeBayet that has been adopted in another task. #185085

I don't see this diff in the task you've mentionned:
- https://github.com/odoo/odoo/pull/185085/files#diff-0ddbec6898f9ebff3f807b92be374389ff3fb45c536daa3b68aab3b9c80e9b18R369-R371

File: addons/portal/views/address_templates.xml (hunk 13 above)
Comment (rahg-odoo) [reply addons/portal/views/address_templates.xml:137]: `id="portal.address_management"` & `id="portal.portal_my_details"` both of these template has `t-call="portal.address_form_fields"`,  but we need to render login field only when called from  `id="portal.portal_my_details"`. 

so yeah, it wont be always true like in case of `id="portal.address_management"` template.

File: addons/portal/controllers/portal.py [hunk 15]
```diff
@@ -160,6 +161,23 @@ def _prepare_portal_layout_values(self):
             'page_name': 'home',
         }
 
+    def _validate_login_field_and_update(self, login, invalid_fields, missing_fields, error_messages):
```
[the remark this answers is not in the document]
Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:164]: @SergeBayet 

We already discussed this point above.

image
then decided to use database constraint instead of using search limit.

File: addons/portal/views/address_templates.xml (hunk 14 above)
Comment (rahg-odoo) [reply addons/portal/views/address_templates.xml:381]: @panv-odoo How did you achieved it and why? can you please explain? Thanks!

Comment (panv-odoo) [reply addons/portal/views/address_templates.xml:381]: > @panv-odoo Could you please explain how you achieved it and the reasoning behind it? Thanks!

We’ve already handled the “Save” and “Save Address” behavior in the PR mentioned above.  
For reference, here’s a video demonstration: [Video](https://www.awesomescreenshot.com/video/38466004?key=750c097071cb2c8b19fbb557fc9d1959)]

We decided to retain the “Save” button on the `my/account` pages, while for address-related forms on `my/address`, we’ve kept the “Save Address” label as it feels more appropriate in that context.

File: addons/portal/views/address_templates.xml (hunk 13 above)
Comment (rahg-odoo) [reply addons/portal/views/address_templates.xml:137]: @SergeBayet Made few changes here, this way there was a limitation that if we are accessing form values through 
`my/account?redirect=/my/addresses` login field was visible, which shouldn't.

File: addons/portal/controllers/portal.py [hunk 16]
```diff
@@ -736,6 +743,37 @@ def _validate_address_values(
...
+        old_login = request.env.user.login
+        if not login:
+            missing_fields.add("login")
+            error_messages.append(_("Some required fields are empty."))
+        elif old_login != login:
+            try:
+                with request.env.cr.savepoint():
+                    request.env.user.write({"login": login})
+                    if error_messages:
+                        request.env.user.write({"login": old_login})
```
Comment (SergeBayet) [inline addons/portal/controllers/portal.py:770]: I still don't understand why you test "error_messages" after the `write`. Can't you test error_messages before?
Something like that:
```
...
elif old_login != login and not error_messages:
...
```

Is this only for this case? (where there are multiple error messages) : 
image

**Review by SergeBayet — COMMENTED** [review]
Just a question and we can go forward.

File: addons/portal/controllers/portal.py (hunk 16 above)
Comment (rahg-odoo) [reply addons/portal/controllers/portal.py:770]: @SergeBayet Yes.

if we choose not to address this point, the errors will appear gradually, one after another, especially when login is involved.

For example, if there are errors in both the email and login fields, only the email error will be shown first. Once that is resolved, the login error will then be displayed.

If we want to handle this without multiple write operations, we should use a search with a limit instead of relying on a database constraint.

Comment (SergeBayet) [reply addons/portal/controllers/portal.py:770]: Ok then. Thanks for the clear explanation.

**Comment by bso-odoo** [comment]
@robodoo delegate=SergeBayet

**Comment by SergeBayet** [comment]
@robodoo r+
